### PR TITLE
Implement runtime polymorphism for function calls

### PR DESCRIPTION
**Pull request by czgdp1807**
This is the initial implementation. Other things like going up the hierarchy ladder if the current class doesn't have a function implemented is yet to be implemented.
The approach is very simple, its just like select-type. Inside each if-else-if ladder block we know which type is present and then the call correct method. So no need to update virtual tables, just compare the types at runtime and then you always know which method is being called from the ASR.

I didn't feel the need to go with the approach of C++/C where they store function pointers inside virtual tables as the same thing is achieved with a simpler logic as described above. Also since we already us that approach in select-type, so I thought to re-use it here as well.

### CODE REVIEW

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -7312,6 +7337,99 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                builder->CreateStore(CreateCallUtil(fn, args, return_var_type0), tmp);
+            }
+            builder->CreateBr(mergeBB);
+
+            start_new_block(elseBB);
+            current_select_type_block_type = nullptr;
+            current_select_type_block_der_type.clear();
+        }
+        start_new_block(mergeBB);
+        tmp = CreateLoad(tmp);
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:7430]: Isn't using the "if" templated function more readable? It seems a lot of repetitive code for an if statement.

**Review by certik — APPROVED** [review]
I think this looks great, thanks for the implementation!

**Comment by certik** [comment]
The C++ approach with pointers might be faster at runtime, but we can do that later if needed.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:7430]: `if` template function is for a single `if-else` block. What I need is an `if-else-if` ladder of size `n`. So, if someone has the time to create a flexible API for creating `if-else-if` ladder of a given size with each block having its own condition then sure I will be happy to use that API. My main priority here was to implement runtime polymorphism, further API abstractions can be done later.

**Comment by czgdp1807** [comment]
> The C++ approach with pointers might be faster at runtime

The only hurdle I feel is that we would need to define all the deferred methods with same signatures for each class and then inside the methods cast to correct child class. Consider the example in, https://github.com/czgdp1807/lfortran/blob/322719707e7356905329e8388556ddac8e14cfa9/integration_tests/class_05.f90. You can see `r2d` accepts an object of type `class(point2d)` and `r3d` defines an object of type `class(point3d)`. In LLVM their definition looks like as follows,

`define float @__module_defs_r2d(%point2d_polymorphic* %this) {`

and 

`define float @__module_defs_r3d(%point3d_polymorphic* %this) {`

So before calling these two I have to cast `class(point) :: ptr` to `class(point2d) :: obj` and `class(point3d) :: obj`. Hence, in that case I already know which function I am calling in a particular `if-else-if` block, so why load it from somewhere. It exactly matches the idea of `select-type`. In fact, any runtime polymorphic call can be converted to `select-type` in an ASR to ASR pass so that we don't have to re-do things.

Also, our approach has only a small overhead of a few integer comparisons before the function call. C++ does it directly via loading the called function from a virtual table it added inside the class. However we don't load the function. So we don't have any in-direction, C++ has it. So it would be hard to say which approach is faster. Plus, I am not sure if performance in runtime polymorphism matters a lot. Let's keep this option for the future.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
Comment (certik) [reply src/libasr/codegen/asr_to_llvm.cpp:7430]: Of course, first get it done, then we iterate if we figure out a better solution.

**Comment by certik** [comment]
I agree. It's not clear to me either which approach is better, we would have to benchmark it.

**Comment by czgdp1807** [comment]
One way to benchmark is implement the exact same logic which I did here in C and then also use C++ virtual and compare. Calling a virtual method in a large loop (say a million) might show which approach has higher overhead. Its a difference of constant times, asymptotically both are same.

**Pull request merged**